### PR TITLE
KAFKA-14979:Incorrect lag was calculated when markPartitionsForTruncation in ReplicaAlterLogDirsThread

### PR DESCRIPTION
**Pull request by hudeqi**
When the partitions of ReplicaFetcherThread finished truncating, the ReplicaAlterLogDirsThread to which these partitions belong needs to be marked truncate. The lag value in the newState (PartitionFetchState) obtained in this process is still the original value (state.lag). If the truncationOffset is smaller than the original state.fetchOffset, then the original lag value is incorrect and needs to be updated. It should be the original lag value plus the difference between the original state.fetchOffset and truncationOffset.

If the original lag value is used incorrectly, then ReplicaAlterLogDirsThread may set the wrong lag value for fetcherLagStats when executing "processFetchRequest". So it might be more reasonable to recalculate a lag value based on new state.fetchOffset.

### CODE REVIEW

**Comment by hudeqi** [comment]
Does anyone care about this issue?

**Comment by hudeqi** [comment]
hi, do you have time to review this issue？ @showuon

**Comment by hudeqi** [comment]
Hi, do you have time to review this? @divijvaidya

File: core/src/main/scala/kafka/server/AbstractFetcherThread.scala [hunk 1]
```diff
@@ -460,8 +460,16 @@ abstract class AbstractFetcherThread(name: String,
     partitionMapLock.lockInterruptibly()
     try {
       Option(partitionStates.stateValue(topicPartition)).foreach { state =>
-        val newState = PartitionFetchState(state.topicId, math.min(truncationOffset, state.fetchOffset),
-          state.lag, state.currentLeaderEpoch, state.delay, state = Truncating,
+        var lag = state.lag
```
Comment (Hangleton) [inline core/src/main/scala/kafka/server/AbstractFetcherThread.scala]: Isn't the lag recalculated on the next fetch iteration ([here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/AbstractFetcherThread.scala#L362))?

Comment (hudeqi) [reply core/src/main/scala/kafka/server/AbstractFetcherThread.scala]: > Isn't the lag recalculated on the next fetch iteration ([here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/AbstractFetcherThread.scala#L362))?

Hi Hangleton, thanks for your review! As you said, under normal circumstances, the lag will be recalculated on the next fetch. However, if an exception occurs before updating the lag (for example, processPartitionData throws an exception due to a disk error) or the return value of logAppendInfoOpt is None, then the lag value will always be or last for a period of time as the previous error value (obviously truncate, lag display is still very large, which seems to deviate greatly from our understanding). Therefore, the modification here is a defensive measure that can effectively avoid such situations. @Hangleton

**Comment by hudeqi** [comment]
> Hi, do you have time to review this? @divijvaidya

Hi, is this free to review? @divijvaidya

File: core/src/main/scala/kafka/server/AbstractFetcherThread.scala (hunk 1 above)
Comment (hudeqi) [reply core/src/main/scala/kafka/server/AbstractFetcherThread.scala]: @Hangleton hi! Are you still following this?

**Comment by hudeqi** [comment]
Hello, please help to review this PR if you have time, thank you! @satishd

**Comment by hudeqi** [comment]
This minor pr is also about replica fetcher thread, please help to review, thanks! @viktorsomogyi

**Comment by hudeqi** [comment]
Hi, could you review this pr? It seems to be similar to KAFKA-15080. @dajac

**Comment by dajac** [comment]
@hudeqi Sure. I am trying to understand the issue. Have you been able to repro it with a unit test?

**Comment by viktorsomogyi** [comment]
@hudeqi I can get to this early next week.

**Comment by viktorsomogyi** [comment]
@Hangleton sorry, I just noticed I somehow removed you unintentionally.

**Comment by hudeqi** [comment]
I am so sorry, I will update the issue and description later, I think the current description is a bit problematic and unclear. @viktorsomogyi @Hangleton

**Pull request closed without merging**